### PR TITLE
Bug/jenkins 37425 run stop toast changes

### PR DESCRIPTION
**Pull request by cliffmeyers**
# Description
- Work in progress; mostly opening this now just for readability of the diff.
- See [JENKINS-XXXXX](https://issues.jenkins-ci.org/browse/JENKINS-XXXXX).

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [ ] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

@jenkinsci/code-reviewers @reviewbybees

### CODE REVIEW

File: blueocean-core-js/src/js/UrlBuilder.js [hunk 1]
```diff
@@ -103,6 +103,11 @@ export const buildRunDetailsUrlFromQueue = (queueItem, isMultiBranch, expectedBu
     // if multi-branch, change the last value of 'pipelines' to 'branches' so it looks like a multibranch REST URL
     if (isMultiBranch) {
         tokens[tokens.length - 4] = 'branches';
+
+        // TODO: compensation until JENKINS-37746 is fixed - encode the branch name since the server only encoded it once
+        if (encodeBranchName) {
+            tokens[tokens.length - 3] = encodeURIComponent(tokens[tokens.length - 3]);
+        }
```
Comment (tfennelly) [inline blueocean-core-js/src/js/UrlBuilder.js]: I don't see any `tokens.length` checks before accessing. Guaranteed to not be needed?

File: blueocean-core-js/src/js/capability.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
+/**
+ * Created by cmeyers on 8/31/16.
+ */
+import es6Promise from 'es6-promise'; es6Promise.polyfill();
```
Comment (tfennelly) [inline blueocean-core-js/src/js/capability.js]: Why is this needed?

File: blueocean-core-js/src/js/capability.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        const storedCapabilities = {};
+
+        Object.keys(map).forEach(className => {
+            const capabilities = map[className];
+            this.store[className] = storedCapabilities[className] = capabilities.classes.slice();
+        });
+
+        return storedCapabilities;
+    }
+}
```
Comment (tfennelly) [inline blueocean-core-js/src/js/capability.js]: Is this another class/module relating to working out capabilities?

File: blueocean-core-js/src/js/components/utils.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+                text: `Started "${name}" #${runId}`,
+                action: 'Open',
+                onActionClick: () => {
+                    if (toastAction) {
+                        toastAction(runDetailsUrl);
+                    }
+                },
+            });
+        });
+};
```
Comment (tfennelly) [inline blueocean-core-js/src/js/components/utils.js]: I don't think I'd ever find this `createRunStartedToast` component here in `blueocean-core-js/src/js/components/utils.js`. Seems like there should be more appropriate place for this, either in a class providing functions for "runs", or in a class providing general purpose "toasts".

**Comment by tfennelly** [comment]
@cliffmeyers hey Cliff, what's the status of this PR ?

**Comment by cliffmeyers** [comment]
@tfennelly this PR is on hold until after JW. Thanks for the feedback; will incorporate it once I get back to this ticket.

**Comment by cliffmeyers** [comment]
Going to open a new PR when this is actually ready.
